src/engine/Filter.cpp

```diff
@@ -232,18 +299,146 @@ vector<RT>* Filter::computeFilterFixedValue(
+void Filter::computeResultFixedValue(
+    ResultTable* result,
+    const std::shared_ptr<const ResultTable> subRes) const {
   LOG(DEBUG) << "Filter result computation..." << endl;
-  shared_ptr<const ResultTable> subRes = _subtree->getResult();
-  result->_nofColumns = subRes->_nofColumns;
-  size_t l = _lhsInd;
-  Id r = _rhsId;
+
+  // interpret the filters right hand side
+  size_t l = _subtree->getVariableColumn(_lhs);
+  Id r;
+  switch (subRes->getResultType(l)) {
+    case ResultTable::ResultType::KB: {
+      std::string rhs = _rhs;
+      if (ad_utility::isXsdValue(rhs)) {
+        rhs = ad_utility::convertValueLiteralToIndexWord(rhs);
+      } else if (ad_utility::isNumeric(_rhs)) {
+        rhs = ad_utility::convertNumericToIndexWord(rhs);
+      }
+      if (_type == SparqlFilter::EQ || _type == SparqlFilter::NE) {
+        if (!getIndex().getVocab().getId(_rhs, &r)) {
+          r = std::numeric_limits<size_t>::max() - 1;
+        }
+      } else if (_type == SparqlFilter::GE) {
+        r = getIndex().getVocab().getValueIdForGE(rhs);
+      } else if (_type == SparqlFilter::GT) {
+        r = getIndex().getVocab().getValueIdForGT(rhs);
+      } else if (_type == SparqlFilter::LT) {
+        r = getIndex().getVocab().getValueIdForLT(rhs);
+      } else if (_type == SparqlFilter::LE) {
+        r = getIndex().getVocab().getValueIdForLE(rhs);
+      }
+      // All other types of filters do not use r and work on _rhs directly
+      break;
+    }
+    case ResultTable::ResultType::VERBATIM:
+      try {
+        r = std::stoull(_rhs);
+      } catch (const std::logic_error& e) {
+        AD_THROW(ad_semsearch::Exception::BAD_QUERY,
+                 "A filter filters on an unsigned integer column, but its "
+                 "right hand side '" +
+                     _rhs + "' could not be parsed as an unsigned integer.");
+      }
+      break;
+    case ResultTable::ResultType::FLOAT:
+      try {
+        float f = std::stof(_rhs);
+        std::memcpy(&r, &f, sizeof(float));
+      } catch (const std::logic_error& e) {
+        AD_THROW(
+            ad_semsearch::Exception::BAD_QUERY,
+            "A filter filters on a float column, but its right hand side '" +
+                _rhs + "' could not be parsed as a float.");
+      }
+      break;
+    case ResultTable::ResultType::TEXT:
+      AD_THROW(ad_semsearch::Exception::NOT_YET_IMPLEMENTED,
+               "Filtering on text type columns is not supported but required "
+               "by filter: " +
+                   asString());
+      break;
+    case ResultTable::ResultType::LOCAL_VOCAB:
+      if (_type == SparqlFilter::EQ || _type == SparqlFilter::NE) {
+        // Find a matching entry in subRes' _localVocab. If _rhs is not in the
+        // _localVocab of subRes r will be equal to  _localVocab.size() and
+        // not match the index of any entry in _localVocab.
+        for (r = 0; r < subRes->_localVocab->size(); r++) {
```

Comment:
This is the GROUP_CONCAT case I was wondering about, right?

If the local vocab is ordered we could optimize this as a binary search, so maybe add a TODO?

Reply:
The problem with keeping the local vocabulary ordered is that every reordering of the vocabulary would also require a rewriting of all local_vocab columns, as the entries in those columns are just offsets into the local_vocab_ vector. Right now the sorting of the local vocabulary would only help us for equality filters on local_vocab columns which should be a rather rare use case, so I believe we would overall loose performance from keeping the local vocabulary sorted.

Reply:
Makes sense, then let's keep it as is